### PR TITLE
Fix #88 function should not have too many lines

### LINKED ISSUE #88: Rule S138: Functions should not have too many lines of code
Implements [RSPEC-138](https://jira.sonarsource.com/browse/RSPEC-138)

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  public FunctionTooBigCheck() {
+    super(UastNode.Kind.FUNCTION);
+  }
+
+  @Override
+  public void visitNode(UastNode node) {
+    FunctionLike from = FunctionLike.from(node);
+    int size = node.lastToken().line - node.firstToken().line;
+    if (size > max) {
+      reportIssue(from.name(), String.format("This function has %s lines of code, which is greater than the %s authorized. Split it into smaller functions.", size, max));
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java]: Bug Bug: A "NullPointerException" could be thrown; "from" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=276&issues=AWL4FL_yqJ_F9Z_shAQr&open=AWL4FL_yqJ_F9Z_shAQr)

[ISSUE_KEY]: AWL4FL_yqJ_F9Z_shAQr

Comment (vilchik-elena) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java]: @benzonico do you plan to fix this issue of NPE?

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  private static final int DEFAULT_MAXIMUM = 120;
+
+  @RuleProperty(
+    key = "max",
+    description = "Maximum authorized lines of code in a function",
+    defaultValue = "" + DEFAULT_MAXIMUM)
+  public int max = DEFAULT_MAXIMUM;
+
+  public FunctionTooBigCheck() {
+    super(UastNode.Kind.FUNCTION);
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java:38]: @benzonico what about function literal? IMO it should be supported by this rule

File: common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
+package FunctionName
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go]: @benzonico `FunctionName`?

File: common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+func foo1() {
+}
+type AI interface {
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go]: @benzonico why this declaration? it seems to not be used (same for other file)

File: common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheckMax5.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package FunctionName
+
+func foo1() {
+}
+type AI interface {
+    //define all methods that you want to override
+    method2()
+}
+
+func foo2() { // Noncompliant {{This function has 6 lines of code, which is greater than the 5 authorized. Split it into smaller functions.}}
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheckMax5.go:10]: @benzonico can we test something in the middle? smaller than max but > 0

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java (hunk 1 above)
Comment (benzonico) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java]: I struggle to see the point : I can add a null check but then how do I actually test it ?

File: common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go (hunk 4 above)
Comment (benzonico) [reply common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go]: I was expecting the method declared to be mapped with kind FUNCTION (and so the functionLike would produce null) but it is not the case, this construction has no uast kind.

Comment (benzonico) [reply common-rule-engine/src/test/files/checks/FunctionTooBigCheck/FunctionTooBigCheck.go]: I think keeping this construction in the test file is a nice safety net for eventual future mapping or uast construction.

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java (hunk 1 above)
Comment (vilchik-elena) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java]: @benzonico don't test it, or at least wont-fix the issue at next

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java (hunk 2 above)
Comment (benzonico) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java:38]: https://jira.sonarsource.com/browse/RSPEC-1188 for me this is another rule.

Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/FunctionTooBigCheck.java:38]: fyi: https://discuss.sonarsource.com/t/rspec-1188-could-be-common-rule/381

**Comment by benzonico** [comment]
@vilchik-elena can you do another round of review on that one ?

**Pull request merged**